### ISSUE TITLE
Unable to change meeting location with latest 3.9.3 update

### ISSUE THREAD

**Issue by marcinyorkpa** [code4recovery/12-step-meeting-list#208]
**Describe the bug**
We are running Wordpress 5.5.3 with TSML version 3.9.3

I'm trying to change a location for an existing meeting.

As soon as I change the Address, the "Update" button becomes disabled, and nothing I do will re-enable it. 

I'm following the steps I've always followed for updating a meeting location:

1. Open the meeting
2. Change the address to the new address
3. Give that address a name
4. Attempt to click "Update"

I don't see any errors in the JavaScript console.

I've tried in both Chrome and Firefox and the behavior is the same

**Expected behavior**
The "Update" button is enabled, and submitting the form saves the meeting update

In addition, the "Preview" button does not work as well. It's clickable, but  nothing happens

**Screenshots**

Initial screenshot of button after pulling up the meeting to edit

![Screenshot 2020-11-01 093341](https://user-images.githubusercontent.com/70951029/97805762-0d281e00-1c26-11eb-8ce3-8eacd2d2f796.png)

The existing location

![Screenshot 2020-11-01 093416](https://user-images.githubusercontent.com/70951029/97805770-1d3ffd80-1c26-11eb-84b2-f8a3a746cd2a.png)

Location after editing it. Note that the map update worked as expected

![Screenshot 2020-11-01 093451](https://user-images.githubusercontent.com/70951029/97805784-3183fa80-1c26-11eb-8f39-8f7644139e13.png)

The Update button after scrolling back up to save changes

![Screenshot 2020-11-01 093508](https://user-images.githubusercontent.com/70951029/97805800-4496ca80-1c26-11eb-83cf-b560a578e87f.png)

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser Chrome and Firefox
 - Version latest versions of Chrome and Firefox

**Comment by tech2serve** [comment]
Please try these steps and see if it works:

1. Open the meeting.
1. Change the Location name to the new location.
1. Change the address to the new address.
1. Attempt to click "Update".

Put another way, reverse the order or your second and third steps.

**Comment by marcinyorkpa** [comment]
thanks @tech2serve 

I've tried as you suggested, but the behavior persists.

The Update button stays submittable when I change the name, but as soon as I clear the Address field, the button is disabled and never reenables

**Comment by tech2serve** [comment]
> but as soon as I clear the Address field, the button is disabled and never reenables

Just confirming, because I don't want to make an unvalidated assumption, that you mean you clear the Address field after changing the Location, and then you enter the new and geocodable address, but the Publish button doesn't enable again?

While I do believe we have a bug, I'm also finding I can work around it on my site by following the steps I outlined. I'm hopeful we can with yours as well. I suggest continuing to try different combinations of clearing both fields and then entering the new information. 

Do you have other addresses to change or is this the only one? Did others work as expected?

**Comment by marcinyorkpa** [comment]
![tsml_address_bug](https://user-images.githubusercontent.com/70951029/97811988-3d82b300-1c4c-11eb-8ae9-8a4c983974ef.gif)

I took a screen capture to show exactly what I'm doing.

I've tried it on multiple meetings, and by creating a brand new meeting and then changing to a different address. I experience the same behavior each time, and no amount of reordering events seems to get the form submittable for me.

thanks again.

**Comment by tech2serve** [comment]
In your screencast, I noticed the notification number for some plugin you have installed increased from 1 to 3:

<img width="56" alt="image" src="https://user-images.githubusercontent.com/31904316/97812643-440f2b80-1c48-11eb-9360-59c868705f53.png">

Can you tell me what that is?

**Comment by marcinyorkpa** [comment]
Yeah, that's LastPass.

I've tried this in a browser window with all extensions disabled, and in Edge where I have no extensions installed, and the behavior is identical across all these different browsers.

Thanks.

**Comment by tech2serve** [comment]
@marcinyorkpa,

I am looking at code to see how we can rework this and make it more robust. But I'm experimenting with different workarounds in order to provide you a solution today. Please try:

1. Open the meeting.
1. Change Address to the new address and tab to let the system geocode the address.
1. Change Location name to the new location, and tab down to Address.
1. Go to the end of Address and delete USA, then tab to let the system geocode a second time.
1. Confirm both Location and Address as are you want them be.
1. Scroll up and see if Publish is enabled. If so, click it.

If this doesn't work for you, then I suspect something particular to your site. If you have a staging site, perhaps load it up with TSML and a set of data, but skip any other plugins you have on your production site. See if the problem exists there.

We've also seen weird problems like this clear up if the website cache is cleared/deleted. This can often be accomplished through your hosting provided c-panel/dashboard.

**Comment by machouinard** [comment]
@marcinyorkpa 

Was this also an issue prior to the WordPress 5.5.3 update, with or without TSML 3.9.3?

One thing you might try is installing and enabling the [Enable jQuery Migrate Helper](https://wordpress.org/plugins/enable-jquery-migrate-helper/) plugin.  It's possible some old jQuery code in another plugin could be causing this.  The plugin will show you warnings about TSML but I believe those are only related to address lookup which appears to be working without issue.

**Comment by marcinyorkpa** [comment]
Thanks so much, @machouinard and @tech2serve 

I was able to get it working with a combination of adding that jquery migrate plugin and following the back-and-forth steps above.

For what it's worth, these are the errors I see from the jquery migration helper:

```
This page generated the following warnings:

    https://york-pa-aa.org/wp-content/plugins/12-step-meeting-list/assets/js/admin.min.js: jQuery.parseJSON requires a valid JSON string
    https://york-pa-aa.org/wp-content/plugins/12-step-meeting-list/assets/js/admin.min.js: jQuery.parseJSON requires a valid JSON string
    https://york-pa-aa.org/wp-content/plugins/12-step-meeting-list/assets/js/admin.min.js: jQuery.parseJSON requires a valid JSON string
```

Now, though, for the address I saved, I'm seeing other weird behavior: On the meeting edit page, the map shows the typical location marker and everything looks fine. But on the published page, the "Directions" box is gone. I've even tried it with different addresses, but none of them will get the "Directions" box back.  Here's a link: https://york-pa-aa.org/meetings/humble-beginnings/ 

I'm mystified. Can you think of anything that would explain this?

Thanks again!

**Comment by tech2serve** [comment]
We're aware and working the jQuery warnings (#168). These issues may be related and I've been working this tonight.

My suspicion regarding why the Directions box is gone is because the location somehow got flagged as an approximate location. Are you familiar with phpMyAdmin?

**Comment by tech2serve** [comment]
@marcinyorkpa,

I am increasingly seeing this issue as NOT related to the jQuery warnings. Nor am I convinced the jQuery warnings are actually directly related to TSML (still working that piece). I have tested meeting editing with the jQuery Migration Helper both activated and deactivated, and the results seem to be the same: I am able to edit meetings, and the only form I see a hiccup consistently with is when I edit the address before editing the name in Location. And in that scenario, I'm able to cause the address to geocode a second time and then update the meeting entry.

My suggestions:
1. If you are familiar with, or have access to, phpMyAdmin, I'd like to get a copy of a cache from your site to confirm my suspicion regarding the directions. So if yes, can we set up a time for me to chat you through this?
1. Continue to look at other plugins you have and try disabling them. I did ask above if you had a staging/demo site you could test with? This is often the quickest and most effective way to confirm another plugin as the problem. 
1. Alternatively, if you export a csv of your meeting data and email it to me (devs@code4recovery.org) I'll bring it up on my dev site and see if I experience the issue.

Thanks.

**Comment by Grateful-Viv** [comment]
I tried to duplicate error. 

Changed address on staging site and pressed update. 

Did not change it.

Then, as suggested, change the name of the location and the street address. Update held.

**Comment by marcinyorkpa** [comment]
@tech2serve thanks again for your reply. I've emailed you with a CSV and some other answers.

I just tried 2 other things:

1) create a brand new meeting, with a brand new location
  - I could not get the Update button to work without having the jquery migrate plugin enabled
  - When I did get the Update button enabled, the Map for this new meeting with new location didn't have the "Directions" bits
2) create a brand new meeting, and associate it with an existing location
  - this time, the direction bits did show up

Not sure if that's helpful or not. Seems like bottom line is that when creating a new location, it gets a bug in its crawl... some value not set correctly or some such thing

**Comment by tech2serve** [comment]
Hey @marcinyorkpa,

Sorry for the delay in getting back to you. I got tied up in some other things. 

In looking at the data you emailed me, it appears to be correct. This causes me to think something in your site is conflicting with our plugin.

Would you send me a screen cap of the location block on the admin side for the Humble Beginnings 9a Sunday meeting (for which the Directions button isn't show up)? 

Also, are you seeing any issues with the Directions in the Meeting Guide app?

**Comment by marcesher** [comment]
Here you go. Thanks!

Not sure if this is helpful, but: I activated the Query Monitor plugin and tried to see if there were any meaningful differences in queries that run for meetings that work vs this busted meeting (and any new meetings with new locations I create), and one thing I see is that in the query that returns information about the location, the existing good meetings don't have any row at all for `is_approximate_location`, whereas this busted meeting does have a row for it. Now, the value is `false` which seems right, so... I dunno. I wonder if it could be a PHP version problem? Our hosting provider is still on PHP 7.2 😬 

As for the meeting guide app... nope, that looks good.

![Screenshot_2020-11-08 Edit Meeting ‹ York Area Intergroup — WordPress(1)](https://user-images.githubusercontent.com/127017/98496819-7dc7cf80-2210-11eb-9871-050e8fdfe8d4.png)

**Comment by marcesher** [comment]
I did a little more debugging and here's what I found.

In the busted meeting, I see this:

`["is_approximate_location"]=> string(5) "false"`

I then added some code into the single meeting template, and I confirmed that the block starting at https://github.com/code4recovery/12-step-meeting-list/blob/4809cafa75bfb2a194e62386e121a7c79020da78/templates/single-meetings.php#L52

`<?php if (!in_array('TC', $meeting->types) && !($meeting->is_approximate_location)) { ?>`

is definitely not getting through due to the `!($meeting->is_approximate_location)` check

I did that by isolating just that part of the conditional and confirming that, in our Wordpress on php 7.2, that block returns `false`, i.e. "not approximate" thinks that it is approximate with the value of `"false"`

In other words, the string "false" is true, and so !(true) is the problem.

Do you think this could be a php version problem?

Thanks again!

**Comment by tech2serve** [comment]
I *think* what we need to figure out is why, at least on your site, the boolean is converted to the string. Or, at least that is what Query Monitor is reporting?  

But I suspect that is what is happening since I believe php of any version will evaluate `$meeting->is_approximate_location === "false"` as true, but I need to confirm that.

**Comment by tech2serve** [comment]
@rs0715 and @marcinyorkpa --

I'm merging discussion regarding the Directions dialogue not displaying into this issue from #217. 

Although I think I know what is going on, I'm unable to figure out so far why it is happening to your sites and not others.

Please reply to this comment with:

1. Hosting provider
1. Wordpress and PHP versions
1. Is `Enable jQuery Migrate Helper` installed and activated?
1. Complete list of other plugins installed. And those that are activated.
1. Whether or not you have a demo/staging site with the same hosting provider and same configuration we can use for testing.

I apologize if some of this is repetitive, but this will help me think of this all in one group,.

Thank you for your help in getting this resolved.

**Comment by tech2serve** [comment]
@rs0715 I’m not seeing an attachment showing other plugins.

**Comment by rs0715** [comment]
[Plugins ‹ — WordPress.pdf](https://github.com/code4recovery/12-step-meeting-list/files/5531601/Plugins.WordPress.pdf)

**Comment by tech2serve** [comment]
@rs0715 Please confirm you've done a test with ALL plugins deactivated EXCEPT TSML and Enable jQuery Migrate Helper. And that during that test, you still did not have the Directions box for meetings that were not location temporarily closed and with a full address?

**Comment by rs0715** [comment]
All plugins disabled - no luck

[Plugins ‹ — WordPress.pdf](https://github.com/code4recovery/12-step-meeting-list/files/5531983/Plugins.WordPress.pdf)
[ss.docx](https://github.com/code4recovery/12-step-meeting-list/files/5531984/ss.docx)

**Comment by marcesher** [comment]
- Hosting provider: iPage
 - Wordpress and PHP versions: WP 5.5.3, PHP 7.2.4
 - Is Enable jQuery Migrate Helper installed and activated? yes
 - Complete list of other plugins installed. And those that are activated: See attached
 - Whether or not you have a demo/staging site with the same hosting provider and same configuration we can use for testing: No

I disabled all plugins except TSML, jquery migrate helper, and google API key, and I still see the behavior described above
![Screenshot_2020-11-12 Plugins ‹ York Area Intergroup — WordPress(1)](https://user-images.githubusercontent.com/127017/99015670-888da780-2523-11eb-809e-a36bea8066e3.png)

**Comment by marcinyorkpa** [comment]
- Hosting provider: iPage
 - Wordpress and PHP versions: WP 5.5.3, PHP 7.2.4
 - Is Enable jQuery Migrate Helper installed and activated? yes
 - Complete list of other plugins installed. And those that are activated: See attached
 - Whether or not you have a demo/staging site with the same hosting provider and same configuration we can use for testing: No

I disabled all plugins except TSML, jquery migrate helper, and google API key, and I still see the behavior described above

![Uploading Screenshot_2020-11-12 Plugins ‹ York Area Intergroup — WordPress(1).png…]()

**Comment by rs0715** [comment]
any update on this?

**Comment by 97eighty** [comment]
@marcinyorkpa @rs0715 
Hi, we're working on reproducing this error. Neither @tech2serve or myself are having luck but we have a few ideas. Is it possible for you to send us an export of your database? We'd like to look closer in to the contents of the post and post_meta tables. Please send to devs@code4recovery.org. Thanks Mark

**Comment by rs0715** [comment]
See attached
[backup_2020-11-14-2013_bgf26dmyftpuploadcom_e9bde7520bb7-db.gz](https://github.com/code4recovery/12-step-meeting-list/files/5541403/backup_2020-11-14-2013_bgf26dmyftpuploadcom_e9bde7520bb7-db.gz)

**Comment by rs0715** [comment]
hello - i restarted my site today along with enabling the api under a different account. Publishing, updating and direction issues seem to be solved. I am wondering though how i can edit and style the meeting this page. I have seen a few sites that have a header or nav bar on that page but i am not sure home to get to it to edit it. Thanks for all your help and info.

**Comment by 97eighty** [comment]
Hi @rs0715 

Good to hear you solved your problem. There are several widgets that allow you to place headers and footers on the meeting list page as well as the meeting detail page. I suggest starting here:

**Comment by rs0715** [comment]
Starting where?

**Comment by tech2serve** [comment]
@marcinyorkpa, It is interesting that @rs0715 was able to solve the issue on his site. Maybe he'll elaborate more on the steps he took to clear the issue?

I am going to work over the the next couple of days on an update to the plugin that *should* make the tracking of approximate locations (which, when true, the directions dialog shouldn't show up) more robust. However, no guarantee that it will solve your issue as we've not been able to duplicate it with our databases.

**Comment by marcinyorkpa** [comment]
@tech2serve perhaps I can help troubleshoot more on my end. Please help me understand something.

In the block that prevents Directions from showing up, looking at this conditional: `!($meeting->is_approximate_location)`

Since:
- That check requires a boolean
- `is_approximate_location` in my case is a string
- the `is_approximate_location` database value is ultimately coming from `wp_postmeta`, with a `meta_key` of `is_approximate_location` and a `meta_value` of "false", which is a `longtext` database field type

My question is: What would cause that to ever get turned into a boolean in the first place? I tried following the code but was unable to determine what would convert the DB string data into a PHP bool. I imagine you will be able to answer that quickly and point me to the exact line of code where that happens.

Based on my admittedly cursory understanding of all this, it sounds like the problem is: the code expects a boolean, and you do too, and in your testing that's what you get, so there's something about either our wordpress install, or database, or something, that is causing a difference in behavior from what you expect, related to datatypes in this case. Or, put shorter and more plainly, the question seems to be: Why isn't this value converting to a bool, as you expect?

Perhaps if I see that, I can troubleshoot further from there.

**Comment by 97eighty** [comment]
@rs0715 

Sorry. Go into the Widgets area and you'll see where you can place headers and footers for those pages.

Mark

**Comment by rs0715** [comment]
a screenshot might be helpful, not sure how your gettingh there, i am using elementor as a page builder but i can use the default also. just not sure how to access the pgs

**Comment by tech2serve** [comment]
@rs0715, I suggest you search for some Wordpress or Elementor tutorials. This isn't really a TSML issue directly. If you're a member of AA, you might also try TIAA Forums.

**Comment by tech2serve** [comment]
> @tech2serve perhaps I can help troubleshoot more on my end. Please help me understand something.
> 
> In the block that prevents Directions from showing up, looking at this conditional: `!($meeting->is_approximate_location)`
> 
> Since:
> 
> * That check requires a boolean
> * `is_approximate_location` in my case is a string
> * the `is_approximate_location` database value is ultimately coming from `wp_postmeta`, with a `meta_key` of `is_approximate_location` and a `meta_value` of "false", which is a `longtext` database field type
> 
> My question is: What would cause that to ever get turned into a boolean in the first place? I tried following the code but was unable to determine what would convert the DB string data into a PHP bool. I imagine you will be able to answer that quickly and point me to the exact line of code where that happens.
> 
> Based on my admittedly cursory understanding of all this, it sounds like the problem is: the code expects a boolean, and you do too, and in your testing that's what you get, so there's something about either our wordpress install, or database, or something, that is causing a difference in behavior from what you expect, related to datatypes in this case. Or, put shorter and more plainly, the question seems to be: Why isn't this value converting to a bool, as you expect?
> 
> Perhaps if I see that, I can troubleshoot further from there.

I'm pretty sure the `is_approximate_location` you're seeing as a string is a result of the `wp_localize_script` call on line 8 of `single_locations.php`. Javascript does indeed appear to be interpreting the value as a string, but in php we handle `is_approximate_location` as a boolean. This is set when you enter a new meeting address (if only city and state, `is_approximate_location` is set to **_true_** and written to the database). The conditional you're referencing above is checking if `$location->is_approximate_location`, a php variable, is **_true_** or **_false_**. The Javascript side of life is not really important in this case.

When I wrote the 3.9 update, I only stored `is_approximate_location` in the database if it was **_true_**, and didn't store anything if it was **_false_**. I did this because I believed that a null value in the database would be read as **_false_**. AFAIK, this seems to be working with nearly all of the 400 or so deployments of v3.9.3. In fact, your website is the only known one right now, and it doesn't seem to be happening with all of your meetings. (This is why @97eighty and I thought if we could get a copy of your database, we could load it and see if we can see what is being handled differently.) 

Is Humble Beginnings  a meeting you first added after your TSML was updated to v3.9.3? And were the other meetings with full addresses in the database before the TSML update? Or, was Humble Beginnings originally just identified by city and state only? And then was changed to be associated with a full address? Are you seeing issues with other meetings? Hopefully this gives you a sense to what we're analyzing. What continues to bewilder us is why this appears to be only happening with the one meeting? My guess is that when we figure out the true cause of the anomaly on your site, we'll have found the real bug. 

I'm grateful for your patience and willingness to work with us.

**Comment by tech2serve** [comment]
> > * the `is_approximate_location` database value is ultimately coming from `wp_postmeta`, with a `meta_key` of `is_approximate_location` and a `meta_value` of "false", which is a `longtext` database field type
> > 

Hey @marcinyorkpa, I missed this above. You are correct that `is_approximate_location` is stored in `wp_postmeta`. Are you saying your database (perhaps you're using _phpMyAdmin_?) shows the field time as `longtext`? Or, how did you determine this? 

I'll need to confirm in my database.

**Comment by marcinyorkpa** [comment]
@tech2serve that is all very helpful in further isolating the problem.

To answer your questions:

This was an existing meeting, but I had changed it to a new location, which I had just added. 

I also tried adding a new meeting, with a different new location, and I saw the same behavior on the meeting screen, with the Directions box not appearing. I then looked in the database and see that it, too, has "is_approximate_location" : "false"

So now it seems the next debugging question is: Why would the code save it if it's false, when you intend it to only save when it's true?

I took a look at the code and, same caveat as before, this is all new to me so I'm pretty much just floundering in the dark here, but when I look at this line:

https://github.com/code4recovery/12-step-meeting-list/blob/master/includes/save.php#L247

It doesn't look like there's any logic that keeps it from saving either a true or a false... it just takes whatever's in the form post and stores it.

I then looked at the geocode code, and then at the form details to see what happens when creating a new Location, and I see that when I add a new location, the `is_approximate_location` hidden field does get set to `false`... see the screenshot

![Screenshot_2020-11-16 Add New Meeting ‹ York Area Intergroup — WordPress](https://user-images.githubusercontent.com/70951029/99338789-e50a0300-2852-11eb-8725-59729e51e202.png)

So based on my limited understanding of the code, it sure looks like the behavior I'm seeing matches what the code is written to do... 1) stick a value in the hidden field, which comes from geocoding, 2) save that value, regardless of whether it's true or false

And then the behavior I see on the meeting page -- with the Directions box not showing up -- also matches what it looks like the code was written to do.

The thing I don't understand is: _how is this working in everyone else's?_ ! 

Thanks again! This is a fun puzzle.

**Comment by marcinyorkpa** [comment]
I also wondered if perhaps the reason it was behaving differently on our site is that I had to add the jquery migrate plugin just to get the "Save" button to work in the first place -- the thing that started this whole thread -- but I just disabled that plugin, tried adding a new meeting with a new location, and confirmed that the `is_approximate_location` hidden field does get populated as `false` in that situation, too (though the save button is still disabled:) )

![Screenshot_2020-11-16 Add New Meeting ‹ York Area Intergroup — WordPress(1)](https://user-images.githubusercontent.com/70951029/99339552-61511600-2854-11eb-977c-d852861b33bf.png)

**Comment by tech2serve** [comment]
Lots of good points here. Thank you.

You're correct that with new meetings, or updates which trigger the `save.php` logic, the `is_approximate_location` is stored, and it should be regardless of the value. However, the database updater logic that executed when users migrated to v3.9 series of the plugin only wrote `is_approximate_location` if the location was determined to be approximate (i.e., city and state only). You can see this in the `database_updates.php` file. 

But this ambiguity that you're alluding to is precisely why I intend to write an update that sets the `is_approximate_location` in the database for every entry, regardless if it is `true` or `false`. Part of my concern is that we don't really know how some of the Wordpress functions are handling the boolean when the initial value may not exist in the database. On the other hand, even your inspections of the final value on the website seems to indicate it is being handled correctly.

I'm further confused by the the fact that this issue continues for you even when you're adding new meetings. So I think it remains worthwhile to explore your database if you can email a copy to us (@97eighty). 

Also, any chance you could ask you hosting provide what version of MySql they are using for your sites?

**Comment by marcinyorkpa** [comment]
Hi @tech2serve . Just emailed you a full database dump. Thanks again for all your work!

**Comment by tech2serve** [comment]
@marcinyorkpa,

Please confirm TSML v3.9.4 resolves the approximate location issue you were experiencing.

**Comment by marcinyorkpa** [comment]
@tech2serve thanks so much for all your work on this and for sticking with it.

I can confirm that the approximate location issue is fixed, and meetings now correctly show the "Directions" button on the meeting screen.

I can also confirm that the "Update button" issue which started all this is still present, and only resolves for me by installing the jquery migrate helper plugin.

Do you have any idea what might be causing that?

thanks!

**Comment by tech2serve** [comment]
Has your Wordpress site updated to version 5.6?

**Comment by Grateful-Viv** [comment]
Yes. No problems detected. Thank you.

**Comment by marcinyorkpa** [comment]
Hi @tech2serve 

I updated our site to WP 5.6 today, and now the location editing behavior is worse than before, in that even with the jquery migrate helper plugin enabled, I can't get the "Update" button to work when I add a new location. 

I'm attaching a few animated gifs to show the behavior, both with and without that plugin installed.

Can you give me a pointer to finding the javascript code responsible for disabling and re-enabling the "Update" button? I tried figuring it out myself, but have so far been unsuccessful. I tried getting Firefox dev tools to break on a change to the button so that I could follow the code, but I'm clearly doing something wrong because I can't get it working. Thanks for any tips to help me debug the javascript.

**with jquery migrate installed and activated**

![tsml_address_bug_with_jquery_migrate_wp_56](https://user-images.githubusercontent.com/70951029/102724207-063f9280-42dc-11eb-8e05-7d401550f93a.gif)

**with jquery migrate deactivated**

![tsml_address_bug_no_jquery_migrate_wp_56](https://user-images.githubusercontent.com/70951029/102724215-18213580-42dc-11eb-8c97-a35f2a95781a.gif)

**Comment by tech2serve** [comment]
Hey @marcinyorkpa,

I appreciate your frustration. Here is what I can tell you from Code for Recovery's work. If you're on Wordpress 5.6 with PHP 7+, and you're using TSML 3.9.4, you should not need Enable jQuery Migrate Helper. This doesn't mean other plugins you have installed aren't conflicting or requiring the Helper. But, from our perspective, with 3.9.4, you don't need it.

It isn't clear from the second gif if the map showed up during the geocode?

Before we proceed with in depth troubleshooting, please test a setup with all other plugins disabled (including Enable jQuery Migrate Helper) and TSML v3.9.4 installed. If the problem persists, reset cache for the server and your browsers. If the problem persists, make sure you haven't made any other modifications to the site (if you're running a child theme, please post the mods here).

If after that, we're still seeing frequent issues with reactivation of the Update button, I'll want to start by getting a detailed description of the setup.

**Comment by marcinyorkpa** [comment]
@tech2serve OK, I think I'm getting somewhere. But where I've gotten is a complete head scratcher to me.

First, I was able to get a copy of my site running locally... both the entire wordpress install, all configs, and a database export. The one thing I can't exactly replicate is whatever the hosting company configures via their php.ini and whatnot. Not sure how important that is right now.

Second, I cannot replicate the problem locally, as you expected. Everything works just fine.

Third, I was able to add a bunch of debugging to isolate where things to awry. So I've learned some things, but I've hit a wall.

In short, things go wrong in the `admin.js`'s `tsml_address` ajax call. I don't know why this is, but that ajax call gets an empty (zero byte) response. I added a `fail()` to see the error, and I see:

```
tsml_address Request Failed: parsererror, SyntaxError: JSON.parse: unexpected end of data at line 1 column 1 of the JSON data
```

which is basically just "empty response"

Now, to try to figure out the empty response, I added some `error_log` debugging into `ajax.php`'s `tsml_admin_ajax_address` function.

What I've observed is that when that first condition trips... the one that just `return array()`, that's when I get the empty response. This makes sense because presumably that's the condition for a new location, i.e. one that doesn't yet exist. Interestingly, locally, this returns `0` and not an empty response.

When the second condition trips, the ajax call works fine, which is why existing locations result in the "Update" button getting reactivated.

So now I've reached a point where I just have no idea why that code would result in an empty response, especially when I can't replicate it locally with the same code and database. 

When I ask myself, "what's different in my real site vs local", the things I think of are:

1. hosting php.ini vs. the default php.ini from the wordpress docker image
2. prod runs ssl whereas local doesn't
3. prod has a big of fiddly code in `wp-config.php` to switch database user, to get around a hosting company queries-per-hour limitation; but I removed that code just to rule it out and it didn't make a difference

I don't see anything obvious in the `debug.log` file on our real site, nothing that clearly indicates an error. Which leads me to suspect that the empty response is happening in the transport layer, or possibly in WP's apparatus that sits on top of the TSML  ajax.php. 🤷 

Even when I just hit the URL directly in a browser, for example `https://york-pa-aa.org/wp-admin/admin-ajax.php?action=tsml_address&formatted_address=705%20E%20Market%20St%2C%20York%2C%20PA%2017403%2C%20USA` -- i.e. not just trying to troubleshoot ajax calls in dev console -- I get the empty response. But not locally.

So: ever see anything this whacky before? I'm going to start googling around for wordpress admin-ajax.php empty response and see if I find anything.

I fully anticipate a big head slap for myself once we figure this out!

Again, I can't thank you enough for how persistent you've been in sticking with me on this. I've worked on open source software for over a decade, and I know how thankless it can seem and how like 95% of the problems you support are problems with other people's setup. I really, really appreciate all your work and help!

**Comment by marcinyorkpa** [comment]
@tech2serve one other thing I see in the `debug.log` but I honestly can't tell if it's related:

```
[21-Dec-2020 21:32:59 UTC] PHP Notice:  Undefined index: types in /wp-content/plugins/12-step-meeting-list/templates/single-locations.php on line 77
[21-Dec-2020 21:32:59 UTC] PHP Notice:  Undefined index: types in /wp-content/plugins/12-step-meeting-list/templates/single-locations.php on line 79
[21-Dec-2020 21:37:07 UTC] PHP Notice:  Undefined index: types in /wp-content/plugins/12-step-meeting-list/includes/shortcodes.php on line 27
[21-Dec-2020 21:37:07 UTC] PHP Notice:  Undefined index: types in /wp-content/plugins/12-step-meeting-list/includes/shortcodes.php on line 40
```

**Comment by 97eighty** [comment]
Hey @tech2serve @marcinyorkpa I've been following this a bit and finally have some cycles to help out. Where do you need me. @marcinyorkpa I know this is probably answered but I picked up something from another support issue. If you wait for a few minutes does the update button work? I'm happy to pull your meetings into an online test build if you think it will help eliminate some of the things to check out.

**Comment by tech2serve** [comment]
@97eighty, a new perspective would be greatly welcomed! 

@marcinyorkpa Which php did you use in your container? 

Frankly, my first guess with the PHP notices is that they're not related, but I'm still confused on why the warnings are getting thrown.

**Comment by marcinyorkpa** [comment]
Knowing practically nothing about php or wordpress, I can't explain this, but:

in ajax.php at line 125, if I change:

`return array()`

to

` wp_send_json("")`

It works just fine

**Comment by 97eighty** [comment]
@marcinyorkpa 7.1 - 7.4

**Comment by 97eighty** [comment]
@marcinyorkpa Pulled your meetings on a fresh Wordpress 5.6 on PHP 7.2. Most plugins disabled. Works as expected. I used your screencast demos to test. Can you post or send a list of your plugins? Go to Tools -> Site Health -> Info - Active Plugins - Copy to Clipboard

**Comment by marcinyorkpa** [comment]
I was able to upgrade to PHP 7.3.2, same behavior.

I tried upgrading to 7.4.10, which my hosting provider supports, but it broke anything with Jetpack, such as Site Health and stats, due to the `ctype` module either not being installed or enabled on 7.4 on my hosting provider, and they either can't or won't change that. 🤷 . That said, meetings and administering meetings still worked, and I confirmed that updating to 7.4 didn't resolve the issue... I still get an empty response from the `tsml_address` ajax call on 7.4

@97eighty I'm not surprised it works as expected for you. It works as expected for me too, locally, using an identical copy of my site (with all plugins) and database. The mystifying thing is that I just cannot figure out what would be different between prod and local, except for the hosting provider's PHP config and SSL.

Why `return array()` would return an empty response, but `wp_send_json("")` would send an empty string and thus work, is beyond me.

Here's the plugin list:

12 Step Meeting List | Version 3.9.4 by Code4Recovery \| Auto-updates enabled
-- | --
12 Step Meeting PDF Generator | Version 0.2.4 by Chris Toews \| Auto-updates enabled
Akismet Anti-Spam | Version 4.1.7 by Automattic \| Auto-updates enabled
API KEY for Google Maps | Version 1.2.0 by GeoDirectory \| Auto-updates enabled
Jetpack by WordPress.com | Version 9.2.1 by Automattic \| Auto-updates enabled
PDF Embedder | Version 4.6.1 by Lever Technology LLC \| Auto-updates enabled
PHP Compatibility Checker | Version 1.5.0 by WP Engine \| Auto-updates disabled
The Events Calendar | Version 5.3.1 by Modern Tribe, Inc. \| Auto-updates enabled
The Events Calendar Category Colors | Version 6.4.11 by Andy Fragen, Barry Hughes \| Auto-updates enabled
The Events Calendar Shortcode & Block | Version 2.4 by Event Calendar Newsletter \| Auto-updates enabled
WordPress Importer | Version 0.7 by wordpressdotorg \| Auto-updates disabled

**Comment by marcinyorkpa** [comment]
@tech2serve I have an observation and a thought:

**Observation**: The ajax.php code that returns when no location is found **seems to work by accident**, not by design. Perhaps (probably) this is because I don't understand something about WP or PHP, but: the return is `return array()` but when you look at what's actually returned when it's working (for me anyways), it's the number `0`. This is what I see in dev tools with the additional debug output I added:

![Screenshot 2020-12-22 111817](https://user-images.githubusercontent.com/70951029/102909766-761e5c00-4447-11eb-9cb4-d78555218685.png)

I see the same thing when I hit the admin-ajax.php endpoint directly... all that's returned is `0`. This is nonintuitive, considering that the code returns an array(). I wonder if this is just a quirk, by accident or design, of how ajax calls work in Wordpress. You can see this in action by changing `return array()` to `return "bill"` or `return True` or anything else... it's always a 0. It is helpful that 0 resolves to False, and thus the corresponding code which deals with the returned data in `admin.js`... `if(data)...` works with that 0 return. 

In the rest of `ajax.php` every other function explicitly calls `wp_send...` , and that jives with all the example code I'm reading online about ajax calls with Wordpress.

**Thought**:  how about changing `return array()` to `wp_send_json("")`

This will guarantee a "Falsey" value is returned and thus `admin.js` will still behave as expected. It also has the benefit of being explicit, and it would be more consistent with the rest of the Ajax functions in this file

Thanks for considering.

**Comment by tech2serve** [comment]
How about doing a PR against the TSML `develop` branch and we can take a look? 

Given this issue you're experiencing seems very isolated and specific to your hosting provider, I'd like to understand the error better before changing the code for 600+ installations. I agree the geocoding -> publish is occasionally a bit flaky, so I think I'd be likely to support revisions that improve. But really to make sure we're not breaking it for others. 

So my suspicion is that your hosting provider has some sort of non-standard setup with their php (who is the provider?) and I'm curious if you've approached them with the issue?

**Comment by marcinyorkpa** [comment]
@tech2serve I spent a long time with tech support (ipage) and they weren't able to figure it out.

I googled the problem, and while I didn't find much, what I did find suggests that the problem with an empty response is usually A) function not properly registered (clearly not the case); B) using `return` instead of echoing and `die`-ing (which `wp_send...` effectively does; C) an error (also clearly not the case)

I definitely understand where you're coming from, being hesitant to change something that's working.

Another option could be to just add a `fail()` handler onto the `getJSON()` call in `admin.js` and have it call `formIsValid()`... again, the point here just being to make the form submittable

**Comment by tech2serve** [comment]
Thanks for the PR. We will try to test it over the next few days. 

Have you tried implementing the change on your production site? Please keep us posted on your experiences there.

**Comment by marcinyorkpa** [comment]
@tech2serve I just updated the PR to have it send `false` rather than an empty string. I think this is more in line with existing behavior, and easier to understand.

Yes, I've tested this on my production site and it works as expected.

Thanks again for considering!
